cheroot/workers/monitor.py

```diff
@@ -0,0 +1,131 @@
+    for each interval, which isn't very responsive; that is, even if you call
+    self.cancel(), you'll have to wait until the sleep() call finishes before
+    the thread stops. To compensate, it defaults to being daemonic, which means
+    it won't delay stopping the whole process.
+
+    :param int interval: time interval between invocation of input function
+    :param function: callable that needs to be invoked
+    :param args: list of arguments to pass to the function
+    :param kwargs: dictionary of keyword args to pass to the function
+    """
+
+    def __init__(self, interval, function, args=None, kwargs=None):
+        """Initialize a BackgroundTask."""
+        super(BackgroundTask, self).__init__()
+        self.interval = interval
+        self.function = function
+        self.args = args if args is not None else []
+        self.kwargs = kwargs if kwargs is not None else {}
+        self.running = False
+        self.daemon = True
+
+    def cancel(self):
+        """Cancel the task."""
+        self.running = False
+
+    def run(self):
+        """Invoke the callable with the given interval."""
+        self.running = True
+        while self.running:
+            time.sleep(self.interval)
+            if not self.running:
+                return
+            # Any error in the calling function will be raised immediatly
+            self.function(*self.args, **self.kwargs)
+
+
+class Monitor:
+    """Monitor a BackgroundTask."""
+
+    callback = None
+    """The function to call at intervals."""
+
+    frequency = 60
+    """The time in seconds between callback runs."""
+
+    thread = None
+    """A :class:`BackgroundTask` thread."""
+    def __init__(self, callback, frequency=60, name=None):
+        """Initialize Monitor."""
+        self.callback = callback
+        self.frequency = frequency
+        self.thread = None
+        self.name = name or self.__class__.__name__
+
+    def start(self):
+        """Start our callback in its own background thread."""
+        if self.frequency <= 0:
+            return
+        if self.thread is not None:
+            return
+        self.thread = BackgroundTask(self.frequency, self.callback)
+        self.thread.setName(self.name)
+        self.thread.start()
+
+    def stop(self):
+        """Stop our callback's background task thread."""
+        if self.thread is None:
+            return
+        if self.thread is not threading.currentThread():
+            self.thread.cancel()
+            if not self.thread.daemon:
+                self.thread.join()
+        self.thread = None
+
+    def graceful(self):
+        """Stop the callback's background task thread and restart it."""
+        self.stop()
+        self.start()
+
+
+class ThreadPoolMonitor(Monitor):
+    """ThreadPoolMonitor for dynamic resizing."""
+
+    def __init__(self, frequency, name=None):
+        """Initialize ThreadPoolMonitor."""
+        self._run = lambda: None
```

Comment:
Not sure if a dummy callable is correct semantically. It looks like just `None` could be better.

Reply:
This callable is bound dynamically after the configure step. 
the monitor run method calls this directly i.e irrespective of configured or not; hence the callable.